### PR TITLE
add lora delete feature

### PR DESCRIPTION
**Pull request by pacman100**
# What does this PR do?
1. Adds `delete_lora` method to the `LoraLoaderMixin`.

### CODE REVIEW

**Review by sayakpaul — APPROVED** [review]
Simple yet so much cool!

File: src/diffusers/utils/peft_utils.py [hunk 1]
```diff
@@ -180,6 +180,14 @@ def set_adapter_layers(model, enabled=True):
                 module.disable_adapters = not enabled
 
 
+def delete_lora_layers(model, adapter_name):
```
Comment (younesbelkada) [inline src/diffusers/utils/peft_utils.py:183]: I think we should raise an error if PEFT version is not greater than 0.6.1

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot! Just left one comment about PEFT versioning

File: src/diffusers/loaders.py [hunk 2]
```diff
@@ -752,6 +753,18 @@ def enable_lora(self):
             raise ValueError("PEFT backend is required for this method.")
         set_adapter_layers(self, enabled=True)
 
+    def delete_lora(self, adapter_name):
```
Comment (younesbelkada) [inline src/diffusers/loaders.py:756]: also what about `delete_adapters` ? To make it in-line with `set_adapters` wdyt @BenjaminBossan @pacman100

Comment (BenjaminBossan) [reply src/diffusers/loaders.py:756]: Good question. AFAICT, there is a bit of a mixture of using the term "lora" or "adapter". I think it depends on whether there is the plan to later enhance this method to support other adapter types. If this is renamed, then `delete_lora_layers` should also be renamed.

When it comes to whether to allow passing a list of names: I see no harm in it.

File: src/diffusers/utils/peft_utils.py (hunk 1 above)
Comment (BenjaminBossan) [reply src/diffusers/utils/peft_utils.py:183]: Agree that a version check would be good to have. I would argue that all checks should be done in one place, i.e. either all inside of `delete_lora` or all here, but not part here part there.

File: src/diffusers/loaders.py (hunk 2 above)
Comment (sayakpaul) [reply src/diffusers/loaders.py:756]: I think `delete_adapters()` is nice. In-line with `set_adapters()` too. 

Could you also include the type-annotation here?

File: src/diffusers/loaders.py [hunk 3]
```diff
@@ -2507,6 +2520,19 @@ def enable_lora(self):
         if hasattr(self, "text_encoder_2"):
             self.enable_lora_for_text_encoder(self.text_encoder_2)
 
+    def delete_lora(self, adapter_name):
+        if not USE_PEFT_BACKEND:
+            raise ValueError("PEFT backend is required for this method.")
+
+        # Enable unet adapters
```
Comment (sayakpaul) [inline src/diffusers/loaders.py:2527]: We're not enabling anything here no?

File: src/diffusers/utils/peft_utils.py (hunk 1 above)
Comment (sayakpaul) [reply src/diffusers/utils/peft_utils.py:183]: Yeah!

File: src/diffusers/loaders.py (hunk 2 above)
Comment (younesbelkada) [reply src/diffusers/loaders.py:756]: Sure ok I'll go with that!

File: src/diffusers/loaders.py (hunk 3 above)
Comment (pacman100) [reply src/diffusers/loaders.py:2527]: But we have enable_lora, disable_lora, unload_lora_weights, load_lora_weights

File: src/diffusers/loaders.py (hunk 2 above)
Comment (sayakpaul) [reply src/diffusers/loaders.py:756]: Could it make sense to allow for deleting multiple LoRAs one-shot like we provide the utility (`set_adapters()`) to handle single adapter loading and multi-adapter loading?  This promotes better feature parity no?

File: tests/lora/test_lora_layers_peft.py [hunk 4]
```diff
@@ -831,6 +831,81 @@ def test_simple_inference_with_text_unet_multi_adapter(self):
...
+            output_adapter_1 = pipe(**inputs, generator=torch.manual_seed(0)).images
+
+            pipe.set_adapters("adapter-2")
+            output_adapter_2 = pipe(**inputs, generator=torch.manual_seed(0)).images
+
+            pipe.set_adapters(["adapter-1", "adapter-2"])
+
+            output_adapter_mixed = pipe(**inputs, generator=torch.manual_seed(0)).images
+
+            # Fuse and unfuse should lead to the same results
```
Comment (sayakpaul) [inline tests/lora/test_lora_layers_peft.py:877]: Irrelevant comment?

File: src/diffusers/loaders.py (hunk 2 above)
Comment (pacman100) [reply src/diffusers/loaders.py:756]: But we have enable_lora, disable_lora, unload_lora_weights, load_lora_weights

**Review by sayakpaul — COMMENTED** [review]
Overall, this looks really cool to me!

Left a question around making `delete_lora()` (or `delete_adapters()`, whatever we decide) accept a list of adapter names.

**Comment by younesbelkada** [comment]
All comments should have been addressed now! Extended also the test to make sure we deal with corner cases

File: src/diffusers/loaders.py [hunk 5]
```diff
@@ -752,6 +753,23 @@ def enable_lora(self):
...
+                The names of the adapter to delete. Can be a single string or a list of strings
+        """
+        if not USE_PEFT_BACKEND:
+            raise ValueError("PEFT backend is required for this method.")
+
+        if isinstance(adapter_names, str):
+            adapter_names = [adapter_names]
+
+        for adapter_name in adapter_names:
+            delete_lora_layers(self, adapter_name)
```
Comment (sayakpaul) [inline src/diffusers/loaders.py:771]: Love!

File: src/diffusers/utils/peft_utils.py (hunk 1 above)
Comment (BenjaminBossan) [reply src/diffusers/utils/peft_utils.py:183]: Should this function also be renamed for consistency to `delete_adapter_layers`?

Comment (younesbelkada) [reply src/diffusers/utils/peft_utils.py:183]: Yes OK

Comment (younesbelkada) [reply src/diffusers/utils/peft_utils.py]: Done in [c54ec0c](https://github.com/huggingface/diffusers/pull/5738/commits/c54ec0cf30ba775cc7782af24a7793b418e9a0ee)

**Review by BenjaminBossan — APPROVED** [review]
Thanks Sourab and Younes, LGTM. Failing code quality checks seem to be unrelated.

**Review by patrickvonplaten — APPROVED** [review]
Clean!

@apolinario can you also give this a look?

**Comment by patrickvonplaten** [comment]
Nice job @pacman100 !
